### ISSUE TITLE
basic_publish of BlockingConnection does not raise exception after network connection goes down

### ISSUE THREAD

**Issue by William-Kong** [pika/pika#400]
We are using pika to connect rabbitmq, the queue is a mirrored queue. I am using the pika in the master branch, which is 0.9.14p0. 

It seems that the basic_publish hangs if the node crashes. And I believe it is busy looping at BlockingConnection._send_method.

The code I have is: 
#!/usr/bin/env python
import pika
import time

connection = pika.BlockingConnection(pika.ConnectionParameters(
        host='example-node'))
channel = connection.channel()

channel.confirm_delivery()
channel.queue_declare(queue='hello')

count = 0
while True:
    try:
        print channel.basic_publish(exchange='', routing_key='hello', body='Hello World!')
        count += 1
        print 'send %d messages' % count
    except:
        break
connection.close()
print 'total send %d messages' % count

After it starts to publish something, I shutdown the node 'example-node', and the script just hangs there. 

I also try to create the connect like: 

connection = pika.BlockingConnection(pika.ConnectionParameters(
        host='example-node',
        connection_attempts=1,
        retry_delay=0,
        socket_timeout=3
 ))

but it doesn't help.

It raises the pika.exceptions.ConnectionClosed, ONLY after I restart the node. 

Am I missing something? Do I have to config the connection/channel in some way? 
Is there even a way to detect that the node is down, or the network connection disappear? Please advice. 

Thank you in advance!

**Comment by vitaly-krugl** [comment]
I would have expected socket timeout to kick in. This is most likely now fixed in pika master via PR #533.

This is a feature of TCP/IP. The default TCP keepalive timeout is typically measured in hours. When the server shuts down gracefully, it makes an attempt to reset open TCP/IP streams, so the remote peers normally recover quickly. However, if a server goes down hard (e.g., kernel panic, hardware failure, pulling the network cable or the power cable), then no one resets the connections, leaving the peers stranded, especially if the peer happens to be reading from the socket at that time. The peer's TCP/IP stack won't discover the disconnect in this scenario until it eventually exercises its keepalive (if supported/enabled) and never hears back, at which point it would close the stream locally.

I recently experienced this with the `requests` package (using default infinite timeout), where a bunch of processes became deadlocked while reading data from a server that malfunctioned. As an experiment, I left them alone and they remained deadlocked for several days until I went in with gdb and forced their socket FDs to close.

See https://delog.wordpress.com/2013/08/16/handling-tcp-keepalive/
